### PR TITLE
Add python subtransaction in plcontainer.

### PR DESCRIPTION
**Pull request by zhangh43**
Now support to use "with plpy.subtransaction():" to open explicit
subtransactions in plcontainer.

### CODE REVIEW

File: src/common/messages/message_subtransaction.h [hunk 1]
```diff
@@ -0,0 +1,35 @@
+/*------------------------------------------------------------------------------
+ *
+ *
+ * Copyright (c) 2017, Pivotal.
```
Comment (paul-guo-) [inline src/common/messages/message_subtransaction.h]: copyright header.

File: src/pyclient/plpy_spi.c [hunk 2]
```diff
@@ -496,19 +507,150 @@ PLy_subtransaction_dealloc(PyObject *subxact UNUSED)
...
+			break;
+		case NO_SUBTRANSACTION_ERROR:
+		case RELEASE_SUBTRANSACTION_ERROR:
+			raise_execution_error("Error receiving subtransaction exit error message");
+			break;
+		case CREATE_SUBTRANSACTION_ERROR:
+			raise_execution_error("Error when beginning subtransaction on QE side");
+			break;
+		default:
+			raise_execution_error("Error receiving unknown error message");
```
Comment (paul-guo-) [inline src/pyclient/plpy_spi.c]: Please specify the result and subtransaction info in the string.

File: src/common/messages/message_subtransaction.h [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+enum SUBTRANSACTION_RETURN_TYPE{
+	NO_SUBTRANSACTION_ERROR = 0,
+	CREATE_SUBTRANSACTION_ERROR,
+	RELEASE_SUBTRANSACTION_ERROR,
+	SUCCESS,
+};
+
+typedef struct plcMsgSubtransaction {
+    base_message_content;
+	char         action;  /* subtransaction action, 'n' for enter and 'x' for exit */
```
Comment (HaozhouWang) [inline src/common/messages/message_subtransaction.h]: indent

File: src/plcontainer.c [hunk 4]
```diff
@@ -40,8 +40,27 @@ static Datum plcontainer_process_result(FunctionCallInfo  fcinfo,
...
+ */
+void
+_PG_init(void)
+{
+	/* Be sure we do initialization only once (should be redundant now) */
+	static bool inited = false;
+
+	if (inited)
+		return;
+
```
Comment (HaozhouWang) [inline src/plcontainer.c]: indent

File: src/common/messages/message_subtransaction.h (hunk 3 above)
Comment (zhangh43) [reply src/common/messages/message_subtransaction.h]: fixed

File: src/plcontainer.c (hunk 4 above)
Comment (zhangh43) [reply src/plcontainer.c]: it's tab indent

**Comment by HaozhouWang** [comment]
+1

**Comment by paul-guo-** [comment]
Maybe have a separate test file for subtransaction. Others look good.

File: src/pyclient/plpy_spi.c (hunk 2 above)
Comment (zhangh43) [reply src/pyclient/plpy_spi.c]: fix it

**Pull request merged**